README.md
[![pub package][pub]][pub-link]
[![BSD License][license-badge]][license-link]
[![PRs Welcome][prs-badge]][prs-link]
[![Last changes][github-changes-badge]][github-changes-link]
[![CI][CI-badge]][CI-link]


[pub]: https://img.shields.io/pub/v/bybit.svg?style=for-the-badge&logo=dart
[pub-link]: https://pub.dev/packages/bybit
[license-badge]: https://img.shields.io/github/license/PimpMyPizza/bybit-dart.svg?style=for-the-badge
[license-link]: https://github.com/PimpMyPizza/bybit-dart/blob/main/LICENSE
[prs-badge]: https://img.shields.io/badge/PRs-welcome-brightgreen.svg?style=for-the-badge
[prs-link]: https://github.com/PimpMyPizza/bybit-dart/issues
[github-changes-badge]: https://img.shields.io/github/last-commit/PimpMyPizza/bybit-dart?style=for-the-badge&logo=git&logoColor=white
[github-changes-link]: https://github.com/PimpMyPizza/bybit-dart/commits/master
[CI-badge]: https://img.shields.io/github/workflow/status/PimpMyPizza/bybit-dart/Dart?logo=github-actions&style=for-the-badge
[CI-link]: https://github.com/PimpMyPizza/bybit-dart/actions


# ByBit

ByBit is a [Dart](https://dart.dev/) library for an easy communication with the [bybit](https://www.bybit.com/) exchange platform [API](https://bybit-exchange.github.io/docs/inverse/#t-introduction). This package allows to make simple REST API calls or to subscribe to several WebSockets channels topics.

## How to use it

### Create a ByBit instance

Note that all the parameters are optional, but you need a valid key and password to access private topics from bybit. You can create your own api-key [on the bybit website](https://www.bybit.com/app/user/api-management).

``` Dart
var bybit = ByBit(
    key: 'yOuRsUpErKey',
    password: 'yOuRsUpErPaSsWoRd',
    logLevel: 'INFO',
    restUrl: 'https://api.bybit.com',
    websocketUrl: 'wss://stream.bytick.com/realtime',
    timeout: 60);
```

Note also that the timeout is given in seconds. If no timeout is given as parameter, no exception will be thrown on timeout.

### Connect to the servers

``` Dart
bybit.connect();
```

### Add periodic REST API call if you want to

Sometimes, you want to get information from the API every x period of time. That's why this library allows one to set which REST API call has to be done periodically, and all the responses from the server are merged into one single stream. Please note [the limit of API calls](https://bybit-exchange.github.io/docs/inverse/#t-ratelimits).

```Dart
bybit.getServerTimePeriodic(period: Duration(seconds: 5));
bybit.getAnnouncementPeriodic(period: Duration(seconds: 5));
bybit.getOpenInterestPeriodic(
    symbol: 'ETHUSD',
    interval: '15min',
    period: Duration(seconds: 2),
    limit: 3);
```


### Subscribe to WebSocket topics

``` Dart
// ...
bybit.subscribeToKlines(symbol: 'ETHUSD', interval: '1');
bybit.subscribeToKlines(symbol: 'BTCUSD', interval: 'D');
bybit.subscribeToOrderBook(depth: 25);
```

### Read the ByBit stream and handle the server response

Note that the `bybit.stream` streams all the data from the WebSockets and periodic REST API calls.

```Dart
StreamBuilder(
    stream: bybit.stream,
    builder: (context, bybitResponse) {
        print('From WebSocket: ' + bybitResponse.data.toString());
        //...
    }
),
//...
```

### You can also make single REST API calls

``` Dart
// ...
FutureBuilder(
    future: bybit.getKLine(symbol: 'BTCUSD', from: 1581231260, interval: 'D'),
    builder: (context, bybitResponse) {
        // Handle the bybit response here
        if (bybitResponse.hasData && bybitResponse.data != null) {
          print('From REST: ' + bybitResponse.data.toString());
          //...
```

## Example

See [the file main.dart](https://github.com/PimpMyPizza/bybit-dart/blob/main/example/lib/main.dart) in the `example/lib/` directory for a simple Dart example. Also, the files [main_flutter_stream.dart](https://github.com/PimpMyPizza/bybit-dart/blob/main/example/lib/main_flutter_stream.dart) and [main_flutter_future.dart](https://github.com/PimpMyPizza/bybit-dart/blob/main/example/lib/main_flutter_future.dart) show examples using `FutureBuilder` and `StreamBuilder`.

## List of functions

See [the documentation](https://pub.dev/documentation/bybit/latest/bybit/ByBit-class.html) for the latest avaiable functions.


CHANGELOG.md
## [2.1.1]

* Minor bug fixes.

## [2.1.0]

* Added API changes until 2022-02-28.
* Minor bug fixes.

## [2.0.0]

* Null safety
* Minor typos corrections and bugfixes

## [1.4.0]

* Removed `int WebSocketTimeout` and `int restTimeout` parameters and replaced them with an `int timeout` parameter.
* Added possibility to set a `receiveWindow` parameter that prevents from replay attacks
* Minor doc changes

## [1.3.0]

* Added the possibility to make periodic REST API calls.
* All API responses from the server (Websockets and periodic REST) are merged into one single stream.
* See README.md for an adapted tutorial
* Minor bugfixed and typos in doc

## [1.2.0]

* Minor changes in the parameters type of the following functions:
  - placeConditionalOrder
  - updateConditionalOrder
  prices and quantities parameters are now of type double/int
* Minor doc changes

## [1.1.0]

* Added new endpoints connections
  - getMarkPriceKLine
  - getOpenInterest
  - getLatestBigDeals
  - getLongShortRatio
* Minor bug-fixes

## [1.0.0]

* More doc, static analysis and other minor changes

## [0.5.0]

* API responses are now automatically mapped to JSON objects
* ByBit object is no more a singleton.
* More doc

## [0.4.0]

* Removed flutter dependency
* Endpoint tests automation
* Documentation of all endpoints parameters

## [0.3.0]

* Enabled Web support
* Added some doc

## [0.2.0]

* Adapted formating
* Added some doc
* Fixed bug with wrong log level
* Web version still has to be debugged

## [0.1.0]

* First version
* WebSocket and HTTPS (REST) communication
* Only a few endpoints were tested


example/lib/main_flutter.dart
import 'package:flutter/material.dart';
import 'package:bybit/bybit.dart';

/// Just a contains that shows the outputs of the ByBit websocket
class ExampleByBitWebSocket extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    ByBit bybit = ByBit.getInstance();
    bybit.subscribeToKlines(symbol: 'ETHUSD', interval: '1');
    bybit.subscribeToKlines(symbol: 'BTCUSD', interval: 'D');
    bybit.subscribeToOrderBook(depth: 25);
    return StreamBuilder(
      stream: bybit.websocket.websocket.stream,
      builder: (context, bybitResponse) {
        // Handle the bybit response here
        if (bybitResponse.hasData && bybitResponse.data != null) {
          //print('From WebSocket: ' + bybitResponse.data.toString());
          return Container(child: Text(bybitResponse.data.toString()));
        } else {
          return Container();
        }
      },
    );
  }
}

/// Just a container that shows the return value from a REST call to bybit
class ExampleByBitREST extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    ByBit bybit = ByBit.getInstance();
    return FutureBuilder(
      //future: bybit.getKLine(symbol: 'BTCUSD', from: 1581231260, interval: 'D'),
      future: bybit.getTickers(),
      builder: (context, bybitResponse) {
        // Handle the bybit response here
        if (bybitResponse.hasData) {
          //print('From REST API: ' + bybitResponse.data.toString());
          return Container(child: Text(bybitResponse.data.toString()));
        } else {
          return Container();
        }
      },
    );
  }
}

class Example extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    // Add a valid key and password if you want to use private topics
    ByBit bybit = ByBit.getInstance(
        logLevel: 'INFO',
        restUrl: 'https://api.bybit.com',
        restTimeout: 3000,
        websocketUrl: 'wss://stream.bytick.com/realtime',
        websocketTimeout: 2000);
    bybit.connect();
    return MaterialApp(
        home: Column(children: <Widget>[
      Container(
          height: 300, color: Colors.blue, child: ExampleByBitWebSocket()),
      Container(height: 300, color: Colors.green, child: ExampleByBitREST())
    ]));
  }
}

void main() {
  runApp(Example());
}
